sys/stdio_nimble/stdio_nimble.c

```diff
@@ -0,0 +1,314 @@
+
+        mutex_unlock(&_rb_mutex);
+        mutex_unlock(&_send_mutex);
+
+        return;
+    }
+
+    int len_read = ringbuffer_get(&_rb, _rb_read_buf, _current_max_read);
+
+    mutex_unlock(&_rb_mutex);
+
+    int rc = 1;
+    if (_send_stdio) {
+        if (_conn_handle != 0 && _val_handle_stdout != 0) {
+
+            struct os_mbuf *om = ble_hs_mbuf_from_flat(_rb_read_buf, len_read);
+
+            if (om != NULL) {
+                _sending_in_progress = true;
+                rc = ble_gattc_indicate_custom(_conn_handle, _val_handle_stdout, om);
+            }
+        }
+    }
+
+    if (rc != 0) {
+        _sending_in_progress = false;
+    }
+
+    mutex_unlock(&_send_mutex);
+}
+
+static void _clear_ringbuffer(void)
+{
+    mutex_lock(&_rb_mutex);
+    ringbuffer_init(&_rb, _rb_mem, RINGBUFFER_SIZE);
+    mutex_unlock(&_rb_mutex);
+}
+
+static int _gap_event_cb(struct ble_gap_event *event, void *arg)
+{
+    (void) arg;
+
+    switch (event->type) {
+
+        case BLE_GAP_EVENT_CONNECT:
+            _clear_ringbuffer();
+            if (event->connect.status) {
+                _send_stdio = false;
+            }
+            _conn_handle = event->connect.conn_handle;
+            ble_att_set_preferred_mtu(MTU_MAX_NIMBLE);
+            ble_gattc_exchange_mtu(_conn_handle, NULL, NULL);
+            break;
+
+        case BLE_GAP_EVENT_DISCONNECT:
+            _send_stdio = false;
+            break;
+
+        case BLE_GAP_EVENT_SUBSCRIBE:
+            if (event->subscribe.attr_handle == _val_handle_stdout) {
+                if (event->subscribe.cur_indicate == 1) {
+                    _send_stdio = true;
+                }
+                else {
+                    _send_stdio = false;
+                }
+            }
+            break;
+
+        case BLE_GAP_EVENT_NOTIFY_TX:
+            if (event->notify_tx.indication == 1) {
+                if (event->notify_tx.status == BLE_HS_EDONE) {
+                    _send_from_ringbuffer(false);
+                }
+                else if (event->notify_tx.status != 0) {
+                    _send_stdio = false;
+                }
+            }
+            break;
+
+        case BLE_GAP_EVENT_MTU:
+            _current_max_read = event->mtu.value - ATT_PKT_HEADER_SIZE;
+            break;
+    }
+
+    return 0;
+}
+
+static int gatt_svr_chr_access_stdin(
+        uint16_t conn_handle, uint16_t attr_handle,
+        struct ble_gatt_access_ctxt *ctxt, void *arg)
+{
+
+    (void) conn_handle;
+    (void) attr_handle;
+    (void) arg;
+
+    uint16_t om_len;
+    om_len = OS_MBUF_PKTLEN(ctxt->om);
+
+    /* read sent data */
+    int rc = ble_hs_mbuf_to_flat(ctxt->om, &_stdin_data, sizeof(_stdin_data), &om_len);
+
+    for (int i = 0; i < om_len; i++) {
+        if (isrpipe_write_one(&_stdin_isrpipe, _stdin_data[i]) == -1) {
+
+            /* sleep for 5 ms to give consumer thread time to consume given data */
+            xtimer_usleep(5000);
```

Comment:
It seems bad to me to block the NimBLE host thread for potentially some time. As mentioned above: I think the way to go is simply to drop data in case the buffer(s) run full!

Reply:
Fixed by dropping characters that don't fit into the ring buffer as you suggested.